add syntax guide to docs

ryzokuken opened the pull request:
fairly minimal, not sure if we need to break this up into the reference section or if we should trim the quickstart, but here's a start we can iterate on

ryzokuken commented:
> I think the syntax guide is a bit in the middle of those right now

Agreed, so should I move it to replace the existing quickstart instead and we can slim it down a bit, starting to move some content to the reference section while we build up the rest?

lucacasonato commented:
I think the best thing to do is if you go through the syntax guide section by section, and for every part decide whether:

1. The content is already in the quick start, in which case drop it.
2. It is more "in-depth" information that should go in reference, in which case we add it there.
3. It is basic information that should go in the quick start, in which case we add it to the quick start.

ryzokuken commented:
I added the declaration section from the guide to the quickstart and made some improvements regarding jargon: https://github.com/lucacasonato/messageformat.dev/pull/1/commits/e9bd4bc2f6d89d8adec9bd45dbc268e9fb59f470

How does this look? Is it too verbose and I should try adapting it instead?

lucacasonato commented on docs/quick-start.md:
```diff
@@ -296,30 +296,19 @@ Selectors also need to be separated by spaces.
+## Input Declarations
 
-`.input` declarations are optional, so a variable that appears without a declaration is assumed to be an external input value.
+An **input** declaration binds a variable to an external input value. It's not required in order to use external input but is a declarative way to verify the presence of (and even the type of) any input.
```
Good! Can you mention that it doesn't just check the type and presence, but can also be used to apply specific formatting to all uses of an input variable (like you do below with `style=percent`). Maybe actually two examples - one showing it just with `:number` with no options, then a paragraph explaining you can also use it to change formatting, and then an example showing how `style=percent` changes the formatting of the variable

lucacasonato commented on docs/reference/functions.md:
````diff
@@ -2,3 +2,14 @@
+Check out {:img src=|image.png|}.
+```
+
+The text `|image.png|` is an example of a _quoted_ literal. It has to be quoted because it includes a '.' character. In general, literals containing non-alphanumeric characters have to be quoted.
````
I think this belongs in literals section?

ryzokuken replied:
done

lucacasonato commented on docs/quick-start.md:
````diff
@@ -298,8 +298,17 @@ values to match against, and can contain multiple variants.
-TODO
+A **local** declaration binds a variable to the value of an expression. `.local` is like `const` in JavaScript. Think of it as a convenient way to reuse the result of an operation without repeating it.
+
+```mf2
+.local $x = {|This is an expression|}
+.local $y = {$now :datetime dateStyle=long}
+```
````
Can you make this an actual runnable example? Right now it is missing the quoted pattern body.

lucacasonato commented on docs/quick-start.md:
```diff
@@ -298,8 +298,17 @@ values to match against, and can contain multiple variants.
 ## Local Declarations
 
-TODO
+A **local** declaration binds a variable to the value of an expression. `.local` is like `const` in JavaScript. Think of it as a convenient way to reuse the result of an operation without repeating it.
```
After the first example we can mention that this is particularly useful for `.match` if you want to use the same formatting across multiple variants of a match statement.

ryzokuken commented:
@lucacasonato marked the PR as ready. I fixed up the examples to look like the rest of the documentation (could wrap them in `mf2-interactive` if needed) and removed the link to the syntax guide. I won't mind removing it altogether but feel like it'll be nice to keep it in if we can to keep track of the parts that didn't make it anywhere else yet. wdyt?

lucacasonato commented:
Yeah sounds good. Can you wrap them in mf2-interactive and then we land it?

ryzokuken commented:
@lucacasonato there goes :D

lucacasonato commented:
All three of these examples are not working quite yet:

![Uploading image.png…]()

ryzokuken commented:
sorry for the oversight, actually the vast majority of the examples didn't work so I made them (in some cases with some tweaks or even improvements) to make them work. Also I realized the usefulness of the interactive widgets, we can visually group message, input data and rendered result very efficiently so thanks again for that.
